### PR TITLE
[SILGen] Unified Entry Point Emission

### PR DESCRIPTION
**Pull request by zachary-kent**
This PR implements unified entry point emission in SILGen so that all top level code is emitted at once. Although this is arguably NFC, I had to modify some tests because all declarations referenced by top-level code are now emitted first during SILGen.

### CODE REVIEW

**Comment by zachary0kent** [comment]
@swift-ci please test

File: lib/SILGen/SILGen.cpp [hunk 1]
```diff
@@ -1963,224 +1927,172 @@ void SILGenModule::visitPoundDiagnosticDecl(PoundDiagnosticDecl *PDD) {
   // Nothing to do for #error/#warning; they've already been emitted.
 }
 
-void SILGenModule::visitTopLevelCodeDecl(TopLevelCodeDecl *td) {
-  assert(TopLevelSGF && "top-level code in a non-main source file!");
+void SILGenModule::emitEntryPoint(SourceFile *SF, SILFunction *TopLevel) {
```
Comment (slavapestov) [inline lib/SILGen/SILGen.cpp:1930]: Could this also go in SILGenTopLevel.cpp?

File: lib/SILGen/SILGenGlobalVariable.cpp [hunk 2]
```diff
@@ -133,6 +133,24 @@ SILGenFunction::emitGlobalVariableRef(SILLocation loc, VarDecl *var,
   return ManagedValue::forLValue(addr);
 }
 
+void SILGenFunction::emitMarkFunctionEscapeForTopLevelCodeGlobals(
```
Comment (slavapestov) [inline lib/SILGen/SILGenGlobalVariable.cpp:136]: Could this go in SILGenTopLevel.cpp?

File: lib/SILGen/SILGenStmt.cpp [hunk 3]
```diff
@@ -908,11 +908,10 @@ void StmtEmitter::visitDeferStmt(DeferStmt *S) {
   // to the top-level code to check initialization of any captured globals.
   FuncDecl *deferDecl = S->getTempDecl();
   auto declCtxKind = deferDecl->getDeclContext()->getContextKind();
-  auto &sgm = SGF.SGM;
-  if (declCtxKind == DeclContextKind::TopLevelCodeDecl && sgm.TopLevelSGF &&
-      sgm.TopLevelSGF->B.hasValidInsertionPoint()) {
-    sgm.emitMarkFunctionEscapeForTopLevelCodeGlobals(
-        S, deferDecl->getCaptureInfo());
+  if (declCtxKind == DeclContextKind::TopLevelCodeDecl &&
```
Comment (slavapestov) [inline lib/SILGen/SILGenStmt.cpp:911]: Does `isa<TopLevelCodeDecl>(deferDecl->getDeclContext())` not compile?

File: lib/SILGen/SILGenTopLevel.cpp [hunk 4]
```diff
@@ -0,0 +1,148 @@
+#include "SILGenTopLevel.h"
```
Comment (slavapestov) [inline lib/SILGen/SILGenTopLevel.cpp:1]: This needs a copyright header as in the other source files

File: lib/SILGen/SILGenTopLevel.cpp [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+#include "SILGenFunction.h"
+#include "Scope.h"
+#include "swift/AST/DiagnosticsSIL.h"
+
+#define DEBUG_TYPE "silgen"
+
+using namespace swift;
+using namespace Lowering;
+
+/// Emit a `makr_function_escape_instruction` into `SGF` if `AFD` captures an
```
Comment (slavapestov) [inline lib/SILGen/SILGenTopLevel.cpp:11]: Typo: `make

File: lib/SILGen/SILGenTopLevel.h [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+#include "SILGen.h"
+#include "swift/AST/ASTVisitor.h"
+#include "swift/AST/SourceFile.h"
+#include "swift/AST/TypeMemberVisitor.h"
+
+namespace swift {
+
+namespace Lowering {
+
+/// Generates a `SILFunction` for toplevel code
```
Comment (slavapestov) [inline lib/SILGen/SILGenTopLevel.h:25]: A bit more detail would be nice; script mode, TopLevelCodeDecl, global variable initializers, etc

Comment (zachary0kent) [reply lib/SILGen/SILGenTopLevel.h]: Makes sense, thanks! Does it look better now?

File: lib/SILGen/SILGen.cpp [hunk 7]
```diff
@@ -1107,6 +1095,14 @@ void SILGenModule::emitFunctionDefinition(SILDeclRef constant, SILFunction *f) {
...
+      auto *File = constant.getFileUnit();
+      // In script mode
+      assert(isa<SourceFile>(File) && "Emitting entry-point of non source file?!");
+      auto *SF = dyn_cast<SourceFile>(File);
+      emitEntryPoint(SF, f);
+      return;
+    }
 
     // TODO: Handle main SourceFile emission (currently done by
     // SourceFileScope).
```
Comment (hamishknight) [inline lib/SILGen/SILGen.cpp:1108]: This comment can be removed now :)

**Review by hamishknight — COMMENTED** [review]
Awesome!

File: lib/SILGen/SILGen.cpp (hunk 7 above)
Comment (zachary0kent) [reply lib/SILGen/SILGen.cpp]: yep!

**Comment by zachary0kent** [comment]
@swift-ci please test

**Review by slavapestov — APPROVED** [review]
This is a nice refactoring, thank you!

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by zachary0kent** [comment]
@swift-ci Please test source compatibility

**Comment by zachary0kent** [comment]
@swift-ci please test

**Review by hamishknight — APPROVED** [review]

**Comment by zachary0kent** [comment]
@swift-ci please test

**Comment by zachary0kent** [comment]
@swift-ci please test

**Pull request merged**